google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/stub/metrics/BuiltinMetricsTracer.java

```diff
@@ -0,0 +1,244 @@
+import com.google.api.gax.tracing.SpanName;
+import com.google.cloud.bigtable.stats.StatsRecorderWrapper;
+import com.google.cloud.bigtable.stats.StatsWrapper;
+import com.google.common.base.Stopwatch;
+import java.util.Map;
+import java.util.concurrent.CancellationException;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicLong;
+import javax.annotation.Nullable;
+import org.threeten.bp.Duration;
+
+/**
+ * A {@link BigtableTracer} that records built-in metrics and publish under the
+ * bigtable.googleapis.com/client namespace
+ */
+class BuiltinMetricsTracer extends BigtableTracer {
+
+  private final StatsRecorderWrapper recorder;
+
+  private final ApiTracerFactory.OperationType operationType;
+  private final SpanName spanName;
+
+  // Operation level metrics
+  private final AtomicBoolean opFinished = new AtomicBoolean();
+  private final Stopwatch operationTimer = Stopwatch.createStarted();
+  private final Stopwatch firstResponsePerOpTimer = Stopwatch.createStarted();
+
+  // Attempt level metrics
+  private int attemptCount = 0;
+  private Stopwatch attemptTimer;
+  private volatile int attempt = 0;
+
+  // Total application latency
+  private final Stopwatch applicationLatencyTimer = Stopwatch.createUnstarted();
+  private final AtomicLong totalApplicationLatency = new AtomicLong(0);
+
+  // Monitored resource labels
+  private String tableId = "undefined";
+  private String zone = "undefined";
+  private String cluster = "undefined";
+
+  // gfe stats
+  private AtomicLong gfeMissingHeaders = new AtomicLong(0);
+
+  BuiltinMetricsTracer(
+      ApiTracerFactory.OperationType operationType,
+      SpanName spanName,
+      Map<String, String> attributes,
+      StatsWrapper statsWrapper,
+      StatsRecorderWrapper recorder) {
+    this.operationType = operationType;
+    this.spanName = spanName;
+    if (recorder != null) {
+      this.recorder = recorder;
+    } else {
+      this.recorder = new StatsRecorderWrapper(operationType, spanName, attributes, statsWrapper);
+    }
+  }
+
+  @Override
+  public Scope inScope() {
+    return new Scope() {
+      @Override
+      public void close() {}
+    };
+  }
+
+  @Override
+  public void operationSucceeded() {
+    recordOperationCompletion(null);
+  }
+
+  @Override
+  public void operationCancelled() {
+    recordOperationCompletion(new CancellationException());
+  }
+
+  @Override
+  public void operationFailed(Throwable error) {
+    recordOperationCompletion(error);
+  }
+
+  @Override
+  public void attemptStarted(int attemptNumber) {
+    attemptStarted(null, attemptNumber);
+  }
+
+  @Override
+  public void attemptStarted(Object request, int attemptNumber) {
+    this.attempt = attemptNumber;
+    attemptCount++;
+    attemptTimer = Stopwatch.createStarted();
+    if (request != null) {
+      this.tableId = Util.extractTableId(request);
+    }
+    if (applicationLatencyTimer.isRunning()) {
```

Comment:
This is inconsistent by using an atomic for applicationLatency I assume that it can be written and read by different threads which is true (gRPC thread for attemptFailed and user thread for request()).
But this would also be true for the applicationLatencyTimer? should there be something to ensure consistency?